## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df

EIN                                               NAME  \
0       10520599                       BLUE KNIGHTS MOTORCYCLE CLUB   
1       10531628             AMERICAN CHESAPEAKE CLUB CHARITABLE TR   
2       10547893                 ST CLOUD PROFESSIONAL FIREFIGHTERS   
3       10553066                     SOUTHSIDE ATHLETIC ASSOCIATION   
4       10556103           GENETIC RESEARCH INSTITUTE OF THE DESERT   
...          ...                                                ...   
34294  996009318              THE LIONS CLUB OF HONOLULU KAMEHAMEHA   
34295  996010315           INTERNATIONAL ASSOCIATION OF LIONS CLUBS   
34296  996012607                                PTA HAWAII CONGRESS   
34297  996015768  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...   
34298  996086871                           WATERHOUSE CHARITABLE TR   

      APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 12 columns]

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df.drop(columns=["EIN","NAME"],inplace=True)
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
app_vcounts=application_df["APPLICATION_TYPE"].value_counts()
app_vcounts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff=500
application_types_to_replace={}
for app in app_vcounts.keys():
  if app_vcounts[app] < cutoff:
    application_types_to_replace[app]="Other"
  else:
    application_types_to_replace[app]=app

# Replace in dataframe
application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].map(application_types_to_replace)

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
class_vcounts=application_df["CLASSIFICATION"].value_counts()
class_vcounts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_vcounts[class_vcounts>1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff=1000
classifications_to_replace={}
for c in class_vcounts.keys():
  if class_vcounts[c] < cutoff:
    classifications_to_replace[c]="Other"
  else:
    classifications_to_replace[c]=c

# Replace in dataframe
application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].map(classifications_to_replace)

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [10]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies=pd.get_dummies(application_df[["APPLICATION_TYPE","AFFILIATION","CLASSIFICATION","USE_CASE","ORGANIZATION","INCOME_AMT"]])
dummies.head()

APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0                       0                     1                     0   
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   

   APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  \
0                    0                    0                    0   
1                    1                    0                    0   
2                    0                    0                    1   
3                    1                    0                    0   
4                    1                    0                    0   

   APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  APPLICATION_TYPE_T8  \
0                    0                    0                    0   
1                    0                    0                    0   
2                    0                    0                    0   
3                    0                    0                    0   
4                    0                    0                    0   

   AFFILIATION_CompanySponsored  ...  ORGANIZATION_Trust  INCOME_AMT_0  \
0                             0  ...                   0             1   
1                             0  ...                   0             0   
2                             1  ...                   0             1   
3                             1  ...                   1             0   
4                             0  ...                   1             0   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  
0                0                  0  
1                0                  0  
2                0                  0  
3                0                  0  
4                0                  0  

[5 rows x 39 columns]

In [11]:
application_prep_df=application_df.drop(columns=["APPLICATION_TYPE","AFFILIATION","CLASSIFICATION","USE_CASE","ORGANIZATION","INCOME_AMT"])
application_prep_df["SPECIAL_CONSIDERATIONS"]=application_prep_df["SPECIAL_CONSIDERATIONS"].map({"N":0,"Y":1})
application_prep_df.head()
for c in dummies.columns:
  application_prep_df[c]=dummies[c]
application_prep_df

STATUS  SPECIAL_CONSIDERATIONS   ASK_AMT  IS_SUCCESSFUL  \
0           1                       0      5000              1   
1           1                       0    108590              1   
2           1                       0      5000              0   
3           1                       0      6692              1   
4           1                       0    142590              1   
...       ...                     ...       ...            ...   
34294       1                       0      5000              0   
34295       1                       0      5000              0   
34296       1                       0      5000              0   
34297       1                       0      5000              1   
34298       1                       0  36500179              0   

       APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0                           0                     1                     0   
1                           0                     0                     0   
2                           0                     0                     0   
3                           0                     0                     0   
4                           0                     0                     0   
...                       ...                   ...                   ...   
34294                       0                     0                     0   
34295                       0                     0                     0   
34296                       0                     0                     0   
34297                       0                     0                     0   
34298                       0                     0                     0   

       APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  ...  \
0                        0                    0                    0  ...   
1                        1                    0                    0  ...   
2                        0                    0                    1  ...   
3                        1                    0                    0  ...   
4                        1                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    0                    1                    0  ...   
34295                    0                    1                    0  ...   
34296                    1                    0                    0  ...   
34297                    0                    0                    1  ...   
34298                    1                    0                    0  ...   

       ORGANIZATION_Trust  INCOME_AMT_0  INCOME_AMT_1-9999  \
0                       0             1                  0   
1                       0             0                  1   
2                       0             1                  0   
3                       1             0                  0   
4                       1             0                  0   
...                   ...           ...                ...   
34294                   0             1                  0   
34295                   0             1                  0   
34296                   0             1                  0   
34297                   0             1                  0   
34298                   0             0                  0   

       INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                           0                         0                   0   
1                           0                         0                   0   
2                           0                         0                   0   
3                           1                         0                   0   
4                           0                         1                   0   
...                       ...                       ...                 ...   
34294                       0                         0           

In [12]:
# Split our preprocessed data into our features and target arrays
y=application_prep_df["IS_SUCCESSFUL"]
X=application_prep_df.drop(columns=["IS_SUCCESSFUL"])
# Split the preprocessed data into a training and testing dataset
X_train,X_test,y_train,y_test=train_test_split(X,y,stratify=y)

In [13]:
X_train

STATUS  SPECIAL_CONSIDERATIONS   ASK_AMT  APPLICATION_TYPE_Other  \
3603        1                       0      5000                       0   
4415        1                       0      5000                       0   
28094       1                       0      5000                       0   
13218       1                       0      5000                       0   
17560       1                       0      5000                       0   
...       ...                     ...       ...                     ...   
681         1                       0      5000                       0   
6856        1                       0      5000                       0   
17696       1                       0      5000                       0   
20817       1                       0     21518                       0   
25512       1                       0  12839230                       0   

       APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
3603                      0                     0                    1   
4415                      0                     0                    1   
28094                     0                     0                    1   
13218                     0                     0                    1   
17560                     0                     0                    1   
...                     ...                   ...                  ...   
681                       0                     0                    1   
6856                      0                     0                    1   
17696                     0                     0                    1   
20817                     0                     0                    1   
25512                     0                     0                    1   

       APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
3603                     0                    0                    0  ...   
4415                     0                    0                    0  ...   
28094                    0                    0                    0  ...   
13218                    0                    0                    0  ...   
17560                    0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
681                      0                    0                    0  ...   
6856                     0                    0                    0  ...   
17696                    0                    0                    0  ...   
20817                    0                    0                    0  ...   
25512                    0                    0                    0  ...   

       ORGANIZATION_Trust  INCOME_AMT_0  INCOME_AMT_1-9999  \
3603                    1             1                  0   
4415                    1             1                  0   
28094                   1             1                  0   
13218                   0             1                  0   
17560                   1             1                  0   
...                   ...           ...                ...   
681                     0             1                  0   
6856                    1             1                  0   
17696                   1             1                  0   
20817                   0             0                  0   
25512                   1             0                  0   

       INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
3603                        0                         0                   0   
4415                        0                         0                   0   
28094                       0                         0                   0   
13218                       0                         0                   0   
17560                       0                         0                   0   
...                       ...                       ...                 .

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()


# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=84, activation="relu", input_dim=42))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=21, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 84)                3612      
                                                                 
 dense_1 (Dense)             (None, 21)                1785      
                                                                 
 dense_2 (Dense)             (None, 1)                 22        
                                                                 
Total params: 5,419
Trainable params: 5,419
Non-trainable params: 0
_________________________________________________________________


In [16]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [17]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)


Epoch 1/100
804/804 [==============================] - 5s 4ms/step - loss: 0.5679 - accuracy: 0.7238
Epoch 2/100
804/804 [==============================] - 6s 7ms/step - loss: 0.5533 - accuracy: 0.7307
Epoch 3/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5505 - accuracy: 0.7331
Epoch 4/100
804/804 [==============================] - 4s 6ms/step - loss: 0.5484 - accuracy: 0.7336
Epoch 5/100
804/804 [==============================] - 5s 7ms/step - loss: 0.5474 - accuracy: 0.7342
Epoch 6/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5464 - accuracy: 0.7346
Epoch 7/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5453 - accuracy: 0.7351
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5455 - accuracy: 0.7365
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5445 - accuracy: 0.7368
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5440 - accura

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


268/268 - 0s - loss: 0.5685 - accuracy: 0.7213 - 474ms/epoch - 2ms/step
Loss: 0.56854647397995, Accuracy: 0.7212827801704407


## The first model has an accuracy of 72.13%, we'll try improving this with different methods.

In [19]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharity.h5")

## Optimization: Trial 1 - Increase epochs

In [23]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_o = tf.keras.models.Sequential()

# First hidden layer
nn_o.add(tf.keras.layers.Dense(units=84, activation="relu", input_dim=42))

# Second hidden layer
nn_o.add(tf.keras.layers.Dense(units=21, activation="relu"))

# Output layer
nn_o.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_o.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 84)                3612      
                                                                 
 dense_7 (Dense)             (None, 21)                1785      
                                                                 
 dense_8 (Dense)             (None, 1)                 22        
                                                                 
Total params: 5,419
Trainable params: 5,419
Non-trainable params: 0
_________________________________________________________________


In [24]:
nn_o.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
fit_model_o = nn_o.fit(X_train_scaled, y_train, epochs=200)



Epoch 1/200
804/804 [==============================] - 3s 2ms/step - loss: 0.5665 - accuracy: 0.7230
Epoch 2/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5536 - accuracy: 0.7314
Epoch 3/200
804/804 [==============================] - 2s 3ms/step - loss: 0.5502 - accuracy: 0.7324
Epoch 4/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5486 - accuracy: 0.7328
Epoch 5/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5477 - accuracy: 0.7341
Epoch 6/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5469 - accuracy: 0.7345
Epoch 7/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5457 - accuracy: 0.7355
Epoch 8/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5452 - accuracy: 0.7349
Epoch 9/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5441 - accuracy: 0.7370
Epoch 10/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5442 - accura

In [22]:
model_loss_o, model_accuracy_o = nn_o.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss_o}, Accuracy: {model_accuracy_o}")

268/268 - 1s - loss: 0.5702 - accuracy: 0.7223 - 634ms/epoch - 2ms/step
Loss: 0.570168137550354, Accuracy: 0.7223323583602905


## The first optimized model has an accuracy of 72.23%, increasing epochs didn't have an effect.



## Optimization: Trial 2 - Increase nodes + try different activation


In [28]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_o2 = tf.keras.models.Sequential()

# First hidden layer
nn_o2.add(tf.keras.layers.Dense(units=168, activation="relu", input_dim=42))

# Second hidden layer
nn_o2.add(tf.keras.layers.Dense(units=42, activation="tanh"))

# Output layer
nn_o2.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_o2.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 168)               7224      
                                                                 
 dense_13 (Dense)            (None, 42)                7098      
                                                                 
 dense_14 (Dense)            (None, 1)                 43        
                                                                 
Total params: 14,365
Trainable params: 14,365
Non-trainable params: 0
_________________________________________________________________


In [29]:
nn_o2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
fit_model_o2 = nn_o2.fit(X_train_scaled, y_train, epochs=100)



Epoch 1/100
804/804 [==============================] - 4s 4ms/step - loss: 0.5663 - accuracy: 0.7243
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5539 - accuracy: 0.7297
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5513 - accuracy: 0.7313
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5487 - accuracy: 0.7324
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5474 - accuracy: 0.7331
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5470 - accuracy: 0.7339
Epoch 7/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5453 - accuracy: 0.7357
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5449 - accuracy: 0.7357
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5441 - accuracy: 0.7354
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5439 - accura

In [30]:
model_loss_o2, model_accuracy_o2 = nn_o2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss_o2}, Accuracy: {model_accuracy_o2}")

268/268 - 0s - loss: 0.5695 - accuracy: 0.7234 - 452ms/epoch - 2ms/step
Loss: 0.569460391998291, Accuracy: 0.7233819365501404


## The second optimized model has an accuracy of 72.33%, increasing the amount of nodes and using tanh instead of relu activation in the hidden layer didn't have an effect.



## Optimization: Trial 3 - Add hidden layers


In [31]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_o3 = tf.keras.models.Sequential()

# First hidden layer
nn_o3.add(tf.keras.layers.Dense(units=82, activation="relu", input_dim=42))

# Second hidden layer
nn_o3.add(tf.keras.layers.Dense(units=42, activation="relu"))

# Third hidden layer
nn_o3.add(tf.keras.layers.Dense(units=21, activation="relu"))

# Fourth hidden layer
nn_o3.add(tf.keras.layers.Dense(units=10, activation="relu"))

# Output layer
nn_o3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_o3.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_15 (Dense)            (None, 82)                3526      
                                                                 
 dense_16 (Dense)            (None, 42)                3486      
                                                                 
 dense_17 (Dense)            (None, 21)                903       
                                                                 
 dense_18 (Dense)            (None, 10)                220       
                                                                 
 dense_19 (Dense)            (None, 1)                 11        
                                                                 
Total params: 8,146
Trainable params: 8,146
Non-trainable params: 0
_________________________________________________________________


In [32]:
nn_o3.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
fit_model_o3 = nn_o3.fit(X_train_scaled, y_train, epochs=100)



Epoch 1/100
804/804 [==============================] - 4s 4ms/step - loss: 0.5675 - accuracy: 0.7241
Epoch 2/100
804/804 [==============================] - 4s 4ms/step - loss: 0.5532 - accuracy: 0.7322
Epoch 3/100
804/804 [==============================] - 6s 7ms/step - loss: 0.5505 - accuracy: 0.7336
Epoch 4/100
804/804 [==============================] - 5s 7ms/step - loss: 0.5483 - accuracy: 0.7347
Epoch 5/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5474 - accuracy: 0.7358
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5465 - accuracy: 0.7369
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5455 - accuracy: 0.7358
Epoch 8/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5450 - accuracy: 0.7355
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5441 - accuracy: 0.7367
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5441 - accura

In [33]:
model_loss_o3, model_accuracy_o3 = nn_o3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss_o3}, Accuracy: {model_accuracy_o3}")

268/268 - 0s - loss: 0.5770 - accuracy: 0.7240 - 451ms/epoch - 2ms/step
Loss: 0.5769902467727661, Accuracy: 0.7239649891853333


## The third optimized model has an accuracy of 72.40%, increasing the amount of layers didn't have an effect.



## Optimization: Trial 4- Combine methods

In [35]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_o_final = tf.keras.models.Sequential()

# First hidden layer
nn_o_final.add(tf.keras.layers.Dense(units=220, activation="relu", input_dim=42))

# Second hidden layer
nn_o_final.add(tf.keras.layers.Dense(units=110, activation="tanh"))

# Third hidden layer
nn_o_final.add(tf.keras.layers.Dense(units=60, activation="relu"))

# Fourth hidden layer
nn_o_final.add(tf.keras.layers.Dense(units=20, activation="tanh"))

# Output layer
nn_o_final.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_o_final.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_25 (Dense)            (None, 220)               9460      
                                                                 
 dense_26 (Dense)            (None, 110)               24310     
                                                                 
 dense_27 (Dense)            (None, 60)                6660      
                                                                 
 dense_28 (Dense)            (None, 20)                1220      
                                                                 
 dense_29 (Dense)            (None, 1)                 21        
                                                                 
Total params: 41,671
Trainable params: 41,671
Non-trainable params: 0
_________________________________________________________________


In [37]:
nn_o_final.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
fit_model_o_final = nn_o_final.fit(X_train_scaled, y_train, epochs=250)



Epoch 1/250
804/804 [==============================] - 6s 6ms/step - loss: 0.5636 - accuracy: 0.7242
Epoch 2/250
804/804 [==============================] - 5s 6ms/step - loss: 0.5534 - accuracy: 0.7310
Epoch 3/250
804/804 [==============================] - 7s 8ms/step - loss: 0.5516 - accuracy: 0.7322
Epoch 4/250
804/804 [==============================] - 6s 8ms/step - loss: 0.5488 - accuracy: 0.7329
Epoch 5/250
804/804 [==============================] - 5s 6ms/step - loss: 0.5482 - accuracy: 0.7346
Epoch 6/250
804/804 [==============================] - 3s 4ms/step - loss: 0.5461 - accuracy: 0.7349
Epoch 7/250
804/804 [==============================] - 3s 3ms/step - loss: 0.5459 - accuracy: 0.7360
Epoch 8/250
804/804 [==============================] - 3s 4ms/step - loss: 0.5449 - accuracy: 0.7354
Epoch 9/250
804/804 [==============================] - 4s 5ms/step - loss: 0.5444 - accuracy: 0.7353
Epoch 10/250
804/804 [==============================] - 3s 4ms/step - loss: 0.5431 - accura

In [38]:
model_loss_o_final, model_accuracy_o_final = nn_o_final.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss_o_final}, Accuracy: {model_accuracy_o_final}")

268/268 - 1s - loss: 0.5779 - accuracy: 0.7207 - 799ms/epoch - 3ms/step
Loss: 0.5778650045394897, Accuracy: 0.7206997275352478


## The accuracy of the model using the most layers, more nodes per layer and even more epochs is 72.07%, basically the same as all others.
## After several tests, this points at the input as the next step. We could try feature engineering, removing features that might not be as important or having more samples could be useful, possibly by using random oversampling.

In [39]:
# Export our optimized model to HDF5 file
nn_o_final.save("AlphabetSoupCharity_optimized.h5")